File: flutter_sdk/lib/src/transport/rest_client.dart
import 'dart:convert';
import 'dart:typed_data';
import 'package:dio/dio.dart';
import '../errors/sdk_error.dart';
import '../crypto/crypto_types.dart';
import 'api_models.dart';

/// Typed REST client for all `rust-e2e-chat-api` endpoints.
///
/// All methods map 1:1 to a server endpoint. Authentication is injected via
/// the [tokenProvider] closure, which is called on every request so that
/// token hot-swaps (via [ChatClient.updateToken]) take effect immediately.
class RestClient {
  final Dio _dio;

  RestClient({
    required String baseUrl,
    required String Function() tokenProvider,
  })  : _dio = Dio(BaseOptions(
          baseUrl: baseUrl,
          connectTimeout: const Duration(seconds: 10),
          receiveTimeout: const Duration(seconds: 30),
          headers: {'Content-Type': 'application/json'},
        )) {
    _dio.interceptors.addAll([
      _AuthInterceptor(tokenProvider),
      _ErrorInterceptor(),
      _RetryInterceptor(dio: _dio),
    ]);
  }

  // ---------------------------------------------------------------------------
  // KDS endpoints
  // ---------------------------------------------------------------------------

  Future<RegisterDeviceResponse> registerDevice(
      String userId, RegisterDeviceRequest request) async {
    final response = await _dio.post(
      '/users/$userId/devices',
      data: jsonEncode(request.toJson()),
    );
    return RegisterDeviceResponse.fromJson(
        response.data as Map<String, dynamic>);
  }

  Future<KeyBundleResponse> getKeyBundle(String userId) async {
    final response = await _dio.get('/users/$userId/key-bundle');
    return KeyBundleResponse.fromJson(response.data as Map<String, dynamic>);
  }

  Future<ReplenishOtpksResponse> replenishOtpks(
      String userId,
      String deviceId,
      List<OneTimePreKey> keys) async {
    final request = ReplenishOtpksRequest(
      oneTimePreKeys: keys
          .map((k) => {
                'id': k.id,
                'public_key': base64.encode(k.publicKey),
              })
          .toList(),
    );
    final response = await _dio.put(
      '/users/$userId/devices/$deviceId/one-time-prekeys',
      data: jsonEncode(request.toJson()),
    );
    return ReplenishOtpksResponse.fromJson(
        response.data as Map<String, dynamic>);
  }

  Future<void> rotateSignedPreKey(
      String userId, String deviceId, SignedPreKeyUpdate update) async {
    await _dio.put(
      '/users/$userId/devices/$deviceId/signed-prekey',
      data: jsonEncode(update.toJson()),
    );
  }

  // ---------------------------------------------------------------------------
  // Conversation endpoints
  // ---------------------------------------------------------------------------

  Future<CreateConversationResponse> createConversation(
      CreateConversationRequest request) async {
    final response = await _dio.post(
      '/conversations',
      data: jsonEncode(request.toJson()),
    );
    return CreateConversationResponse.fromJson(
        response.data as Map<String, dynamic>);
  }

  Future<SendMessageResponse> sendMessage(
      String conversationId, SendMessageRequest request) async {
    final response = await _dio.post(
      '/conversations/$conversationId/messages',
      data: jsonEncode(request.toJson()),
    );
    return SendMessageResponse.fromJson(response.data as Map<String, dynamic>);
  }

  Future<GetMessagesResponse> getMessages(
    String conversationId, {
    int limit = 50,
    int? beforeSeq,
  }) async {
    final queryParams = {
      'limit': limit,
      if (beforeSeq != null) 'before_seq': beforeSeq,
    };
    final response = await _dio.get(
      '/conversations/$conversationId/messages',
      queryParameters: queryParams,
    );
    return GetMessagesResponse.fromJson(response.data as Map<String, dynamic>);
  }

  // ---------------------------------------------------------------------------
  // Group endpoints
  // ---------------------------------------------------------------------------

  Future<CreateGroupResponse> createGroup(
      CreateGroupRequest request) async {
    final response = await _dio.post(
      '/groups',
      data: jsonEncode(request.toJson()),
    );
    return CreateGroupResponse.fromJson(
        response.data as Map<String, dynamic>);
  }

  Future<SendMessageResponse> sendGroupMessage(
      String conversationId, SendMessageRequest request) async {
    final response = await _dio.post(
      '/groups/$conversationId/messages',
      data: jsonEncode(request.toJson()),
    );
    return SendMessageResponse.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> addGroupMember(
      String conversationId, String userId, String deviceId) async {
    await _dio.post(
      '/groups/$conversationId/members',
      data: jsonEncode({'user_id': userId, 'device_id': deviceId}),
    );
  }

  Future<void> removeGroupMember(
      String conversationId, String userId) async {
    await _dio.delete('/groups/$conversationId/members/$userId');
  }

  Future<void> distributeGroupSenderKey(
      String conversationId, List<SkdmRecipient> recipients) async {
    await _dio.post(
      '/groups/$conversationId/sender-key-distribution',
      data: jsonEncode({'recipients': recipients.map((r) => r.toJson()).toList()}),
    );
  }

  // ---------------------------------------------------------------------------
  // Attachment endpoints
  // ---------------------------------------------------------------------------

  Future<UploadResponse> uploadAttachment(
    Uint8List bytes,
    String filename,
    String contentType, {
    void Function(int sent, int total)? onProgress,
  }) async {
    if (bytes.length > 100 * 1024 * 1024) {
      throw FileTooLargeError(sizeBytes: bytes.length);
    }

    final formData = FormData.fromMap({
      'file': MultipartFile.fromBytes(
        bytes,
        filename: filename,
        contentType: DioMediaType.parse(contentType),
      ),
    });

    final response = await _dio.post(
      '/attachments',
      data: formData,
      onSendProgress: onProgress,
    );
    return UploadResponse.fromJson(response.data as Map<String, dynamic>);
  }
}

// ---------------------------------------------------------------------------
// Interceptors
// ---------------------------------------------------------------------------

class _AuthInterceptor extends Interceptor {
  final String Function() _tokenProvider;

  _AuthInterceptor(this._tokenProvider);

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) {
    options.headers['Authorization'] = 'Bearer ${_tokenProvider()}';
    handler.next(options);
  }
}

class _ErrorInterceptor extends Interceptor {
  @override
  void onError(DioException err, ErrorInterceptorHandler handler) {
    final response = err.response;
    if (response != null) {
      String? errorCode;
      String message = 'Request failed';
      try {
        final data = response.data;
        if (data is Map<String, dynamic>) {
          errorCode = data['error_code'] as String?;
          message = data['message'] as String? ?? message;
        }
      } catch (_) {}

      final sdkError = sdkErrorFromApiResponse(
        statusCode: response.statusCode ?? 0,
        errorCode: errorCode,
        message: message,
      );
      handler.reject(
        DioException(
          requestOptions: err.requestOptions,
          error: sdkError,
          type: DioExceptionType.badResponse,
          response: response,
          message: sdkError.toString(),
        ),
      );
    } else {
      handler.reject(
        DioException(
          requestOptions: err.requestOptions,
          error: NetworkError(message: err.message ?? 'Connection failed'),
          type: err.type,
          message: err.message,
        ),
      );
    }
  }
}

class _RetryInterceptor extends Interceptor {
  final Dio dio;
  static const _maxRetries = 3;
  static const _retryDelay = Duration(milliseconds: 500);

  _RetryInterceptor({required this.dio});

  @override
  Future<void> onError(
      DioException err, ErrorInterceptorHandler handler) async {
    final response = err.response;
    final retryCount =
        (err.requestOptions.extra['retryCount'] as int?) ?? 0;

    final shouldRetry = retryCount < _maxRetries &&
        (response == null ||
            response.statusCode == 503 ||
            response.statusCode == 500 ||
            err.type == DioExceptionType.connectionTimeout ||
            err.type == DioExceptionType.receiveTimeout);

    if (shouldRetry) {
      await Future.delayed(_retryDelay * (retryCount + 1));
      final options = err.requestOptions;
      options.extra['retryCount'] = retryCount + 1;
      try {
        final retryResponse = await dio.fetch(options);
        handler.resolve(retryResponse);
        return;
      } catch (e) {
        handler.next(err);
        return;
      }
    }

    handler.next(err);
  }
}

File: web_sdk/src/transport/rest-client.ts
// src/transport/rest-client.ts

import { SdkError, SdkErrorCode } from '../errors/sdk-error.js';
import { withRetry } from '../utils/retry.js';
import type {
  KeyBundleRequest,
  RegisterDeviceResponse,
  KeyBundleResponse,
  OtpkEntry,
  ReplenishResponse,
  SignedPreKeyUpdate,
  CreateConversationRequest,
  CreateConversationResponse,
  MessageEnvelopeRequest,
  SendMessageResponse,
  GetMessagesResponse,
  CreateGroupRequest,
  CreateGroupResponse,
  SkdmRecipient,
  UploadResponse,
  ApiErrorBody,
} from './api-types.js';

/**
 * Typed fetch-based REST client.
 * Uses a `getToken()` closure to support hot-swap via `updateToken`.
 * All key bytes are base64url-encoded in JSON.
 */
export class RestClient {
  constructor(
    private readonly baseUrl: string,
    private readonly getToken: () => string,
  ) {}

  // --- Device ---

  /**
   * Register a new device and receive a DeviceId.
   */
  async registerDevice(userId: string, bundle: KeyBundleRequest): Promise<RegisterDeviceResponse> {
    return this.post<RegisterDeviceResponse>(`/users/${userId}/devices`, bundle);
  }

  /**
   * Fetch a user's public key bundle for X3DH.
   */
  async getKeyBundle(userId: string): Promise<KeyBundleResponse> {
    return this.get<KeyBundleResponse>(`/users/${userId}/key-bundle`);
  }

  /**
   * Upload new one-time prekeys for replenishment.
   */
  async replenishOtpks(
    userId: string,
    deviceId: string,
    keys: OtpkEntry[],
  ): Promise<ReplenishResponse> {
    return this.put<ReplenishResponse>(`/users/${userId}/devices/${deviceId}/one-time-prekeys`, { keys });
  }

  /**
   * Upload a new signed prekey after rotation.
   */
  async rotateSignedPreKey(
    userId: string,
    deviceId: string,
    update: SignedPreKeyUpdate,
  ): Promise<void> {
    await this.put<void>(`/users/${userId}/devices/${deviceId}/signed-prekey`, update);
  }

  // --- Conversations ---

  /**
   * Create a new 1:1 conversation with X3DH init envelope.
   */
  async createConversation(
    req: CreateConversationRequest,
  ): Promise<CreateConversationResponse> {
    return this.post<CreateConversationResponse>('/conversations', req);
  }

  /**
   * Send a Double Ratchet message envelope.
   */
  async sendMessage(
    conversationId: string,
    req: MessageEnvelopeRequest,
  ): Promise<SendMessageResponse> {
    return this.post<SendMessageResponse>(`/conversations/${conversationId}/messages`, req);
  }

  /**
   * Fetch message history for a conversation.
   */
  async getMessages(
    conversationId: string,
    params?: { limit?: number; beforeSeq?: number },
  ): Promise<GetMessagesResponse> {
    const qs = new URLSearchParams();
    if (params?.limit !== undefined) qs.set('limit', String(params.limit));
    if (params?.beforeSeq !== undefined) qs.set('before_seq', String(params.beforeSeq));
    const query = qs.toString() ? `?${qs.toString()}` : '';
    return this.get<GetMessagesResponse>(`/conversations/${conversationId}/messages${query}`);
  }

  // --- Groups ---

  /**
   * Create a new group conversation.
   */
  async createGroup(req: CreateGroupRequest): Promise<CreateGroupResponse> {
    return this.post<CreateGroupResponse>('/groups', req);
  }

  /**
   * Add a member to a group.
   */
  async addGroupMember(
    conversationId: string,
    userId: string,
    deviceId: string,
  ): Promise<void> {
    await this.post<void>(`/groups/${conversationId}/members`, { user_id: userId, device_id: deviceId });
  }

  /**
   * Remove a member from a group.
   */
  async removeGroupMember(conversationId: string, userId: string): Promise<void> {
    await this.delete(`/groups/${conversationId}/members/${userId}`);
  }

  /**
   * Distribute encrypted Sender Key Distribution Messages.
   */
  async distributeGroupSenderKey(
    conversationId: string,
    recipients: SkdmRecipient[],
  ): Promise<void> {
    await this.post<void>(`/groups/${conversationId}/sender-keys`, { recipients });
  }

  // --- Attachments ---

  /**
   * Upload an encrypted attachment.
   */
  async uploadAttachment(
    data: FormData,
    onProgress?: (sent: number, total: number) => void,
  ): Promise<UploadResponse> {
    // Use XHR for progress tracking in browser; fall back to fetch otherwise
    if (typeof XMLHttpRequest !== 'undefined' && onProgress) {
      return this.xhrUpload(data, onProgress);
    }
    return this.post<UploadResponse>('/attachments', data, { raw: true });
  }

  // --- HTTP helpers ---

  private async get<T>(path: string): Promise<T> {
    return this.request<T>('GET', path);
  }

  private async post<T>(path: string, body?: unknown, opts?: { raw?: boolean }): Promise<T> {
    return this.request<T>('POST', path, body, opts);
  }

  private async put<T>(path: string, body?: unknown): Promise<T> {
    return this.request<T>('PUT', path, body);
  }

  private async delete(path: string): Promise<void> {
    await this.request<void>('DELETE', path);
  }

  private async request<T>(
    method: string,
    path: string,
    body?: unknown,
    opts?: { raw?: boolean },
  ): Promise<T> {
    const url = `${this.baseUrl}${path}`;

    const execute = async (): Promise<T> => {
      const headers: Record<string, string> = {
        Authorization: `Bearer ${this.getToken()}`,
      };

      let bodyInit: BodyInit | undefined;
      if (body !== undefined) {
        if (opts?.raw && body instanceof FormData) {
          bodyInit = body;
        } else {
          headers['Content-Type'] = 'application/json';
          bodyInit = JSON.stringify(body);
        }
      }

      let response: Response;
      try {
        response = await fetch(url, { method, headers, ...(bodyInit !== undefined ? { body: bodyInit } : {}) });
      } catch (err) {
        throw SdkError.network(`Network request failed: ${method} ${path}`, err);
      }

      if (!response.ok) {
        let errorBody: ApiErrorBody = {};
        try {
          errorBody = (await response.json()) as ApiErrorBody;
        } catch {
          // ignore parse failure
        }
        throw SdkError.fromApiResponse(response.status, errorBody);
      }

      if (response.status === 204 || response.headers.get('content-length') === '0') {
        return undefined as T;
      }

      try {
        return (await response.json()) as T;
      } catch (err) {
        throw new SdkError(SdkErrorCode.NETWORK_ERROR, 'Failed to parse response JSON', { cause: err });
      }
    };

    // Retry on 503 and network errors, not on 4xx
    return withRetry(execute, { attempts: 3, backoffMs: 500 });
  }

  private xhrUpload(data: FormData, onProgress: (sent: number, total: number) => void): Promise<UploadResponse> {
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.open('POST', `${this.baseUrl}/attachments`);
      xhr.setRequestHeader('Authorization', `Bearer ${this.getToken()}`);

      xhr.upload.onprogress = (e) => {
        if (e.lengthComputable) onProgress(e.loaded, e.total);
      };

      xhr.onload = () => {
        if (xhr.status >= 200 && xhr.status < 300) {
          try {
            resolve(JSON.parse(xhr.responseText) as UploadResponse);
          } catch (err) {
            reject(new SdkError(SdkErrorCode.NETWORK_ERROR, 'Failed to parse upload response', { cause: err }));
          }
        } else {
          let body: ApiErrorBody = {};
          try { body = JSON.parse(xhr.responseText) as ApiErrorBody; } catch { /* ignore */ }
          reject(SdkError.fromApiResponse(xhr.status, body));
        }
      };

      xhr.onerror = () => reject(SdkError.network('XHR upload failed'));
      xhr.send(data);
    });
  }
}
